### PR TITLE
Lf 4260b patch and delete product

### PR DESCRIPTION
**Pull request by Duncan-Brain**
**Description**

Patch and delete product check. (sorry for all the merge commits, I should have waited to add the PR and then rebase before posting it or something)

Changes:
- Adds check for moisture and compounds
- Prevents having wrong task type (ie cleaning) with wrong product details (ie soil amendmnet)
- Prevents switching product type for now to prevent extra orphan product details
- require name for product
- require type for product 

Not done (migration):
- add type as required for database check (consider adding to LF-4246 with other migrations)

Code smell:
- type should not be task type but product type id or something

Jira link: [LF-4260](https://lite-farm.atlassian.net/browse/LF-4260)

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Passes test case
- [ ] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

[LF-4260]: https://lite-farm.atlassian.net/browse/LF-4260?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (kathyavini) [inline packages/api/src/middleware/validation/checkProductValidity.js:105]:
```suggestion
      const nonModifiableAssets = Object.values(taskTypeProductMap).filter((productType) => {
  return productType !== taskTypeProductMap[type];
});
```

If you use `Object.values` you don't have to define both the map and the array of strings and can delete the `typesOfProduct` array. It wouldn't make sense to make the content of the array + map ever diverge, right? 🤔

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (kathyavini) [inline packages/api/src/middleware/validation/checkProductValidity.js:72]: Did you also want to put the `<= 100` check on `moisture_content` here, or let the db handle it?

**Review by kathyavini — COMMENTED** [review]
These looks like good sensible checks to me!

I just saw one below with faulty logic (the negative `moisture_content`) and I couldn't get the "must not have other product type details" check to trigger but I might not be understanding it.

I'm not sure what the test situation is here -- I'm getting a lot of failures now locally with these changes, but I wouldn't mind waiting until the other one is merged and this is based to integration to double-check on that.

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (Duncan-Brain) [reply packages/api/src/middleware/validation/checkProductValidity.js:72]: I think I try to let the natural db errors handle it... to not have extra code.

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (Duncan-Brain) [reply packages/api/src/middleware/validation/checkProductValidity.js]: Yah that seems way better!

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (Duncan-Brain) [reply packages/api/src/middleware/validation/checkProductValidity.js]: yup that was silly of me

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (Duncan-Brain) [reply packages/api/src/middleware/validation/checkProductValidity.js:113]: Yeah I copied the name to be consistent -- if I change it then I think I should change them all so it is searchable / fixable.

That is a strange error, you are right that since the map goes to null for the other products it does nothing for them -- I can delete that. But what you should be seeing is still a 400 from a different check "Must not have additional properties". So I am not quite sure what happened there.

Screenshot 2024-07-10 at 10 56 45 AM

Comment (kathyavini) [reply packages/api/src/middleware/validation/checkProductValidity.js:113]: Is it because I was doing a PATCH and you are doing a POST? Sorry I didn't include the whole postman UI just to save space but it was a PATCH to that `product_id` above!

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (SayakaOno) [inline packages/api/src/middleware/validation/checkProductValidity.js:107]: Why wasn't this formatted? Interesting...

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (SayakaOno) [inline packages/api/src/middleware/validation/checkProductValidity.js:40]: Can this be removed?

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (Duncan-Brain) [reply packages/api/src/middleware/validation/checkProductValidity.js]: I accepted Github commit suggestion -- no pre-commit. :(

Comment (SayakaOno) [reply packages/api/src/middleware/validation/checkProductValidity.js]: Oh I see!

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (Duncan-Brain) [reply packages/api/src/middleware/validation/checkProductValidity.js:126]: Okay I think I found that at least for this case it should have had remove additional properties in the baseController. It should now succeed with extra properties that are not part of nonModifiableAssets.

File: packages/api/tests/mock.factories.js
Comment (SayakaOno) [inline packages/api/tests/mock.factories.js:1154]: It's hard to imagine how this function will be utilized... you're thinking of the possibility of updating `...defaultData`(line 1157), right? 😂

File: packages/api/tests/product.test.js
Comment (SayakaOno) [inline packages/api/tests/product.test.js:410]: I think you forgot to remove this.

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (SayakaOno) [inline packages/api/src/middleware/validation/checkProductValidity.js:51]: Is this different from the DB constraint `elemental_unit_check`?

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (Duncan-Brain) [reply packages/api/src/middleware/validation/checkProductValidity.js:112]: @SayakaOno this one is on purpose actually! 😂  These checks aren't needed as they are on the model -- these checks prevent an error on Line 133 for name uniqueness check.

File: packages/api/tests/mock.factories.js
Comment (Duncan-Brain) [reply packages/api/tests/mock.factories.js:1154]: No product details (soil_amendment_product) are required but the shape of the object is. `{}` lol

If you want to test npk on backend or something you could update this to have default values or pass them to override.

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (Duncan-Brain) [reply packages/api/src/middleware/validation/checkProductValidity.js:51]: I think you are right this is a duplicate .. I just moved it here it isn't a new check.

But I will try removing that.

File: packages/api/src/middleware/validation/checkProductValidity.js
Comment (SayakaOno) [reply packages/api/src/middleware/validation/checkProductValidity.js:107]: Sorry 🤣 Now I understand the code 😂

**Review by SayakaOno — APPROVED** [review]
Looks good to me!

**Review by kathyavini — APPROVED** [review]
Looks great!

**Pull request merged**
